File: simpleq.js
// simpleq
// A simple queuing system

// builtin
var path = require('path');

// vendor
var redis = require('redis'),
  reval = require('redis-eval');

// local
var Listener = require('./listener');

var safepullpipe_filename = path.join(__dirname, '/scripts/safepullpipe.lua');

// -- Master Type: Q --
// The master type, a queue of course
function Q(redisClient, key) {
  // handle forgetting a 'new'
  if (!(this instanceof Q)) {
    return new Q(redisClient, key);
  }

  this._redis = redisClient;
  this._key = key;
}

// Clone this Q by creating a new redis connection
Q.prototype.clone = function clone() {
  var rclone = redis.createClient(this._redis.port, this._redis.host, this._redis.options);
  return new Q(rclone, this._key);
};

// Push an element onto the queue
// Returns length of queue
Q.prototype.push = function push(el, cb) {
  this._redis.lpush(this._key, el, cb);
};

// Pop an element off the queue
// Returns element or nil
Q.prototype.pop = function pop(cb) {
  this._redis.rpop(this._key, cb);
};

// Block and Pop an element off the queue
// Returns element
Q.prototype.bpop = function bpop(timeout, cb) {
  if (cb === undefined) {
    cb = timeout;
    timeout = 0;
  }

  this._redis.brpop(this._key, timeout, function (err, result) {
    cb(err, result && result.length === 2 && result[1]);
  });
};

// Pull n element out of a queue. (The highest/oldest one if elements are repeated)
// Returns the number of elements removed (0 or 1)
Q.prototype.pull = function pull(el, cb) {
  this._redis.lrem(this._key, -1, el, cb);
};

// Pull an element out of a queue and put in another queue atomically.
// Return the number of elements in the other queue
// Note: This will insert an element into the second queue regardless of whether it exists in the first queue
Q.prototype.pullpipe = function pullpipe(otherQ, el, cb) {
  this._redis.multi()
    .lrem(this._key, -1, el)
    .lpush(otherQ._key, el)
    .exec(function (err, replies) {
      if (err) {
        return cb(err);
      }
      cb(undefined, replies && replies.length > 1 && replies[1]);
    });
};

// Safely pull an element out of a queue and put in another atomically.
// If the element does not exist in the queue, it is not inserted in the second queue
// Returns 0 for non-existance in first queue, or length of second queue
Q.prototype.spullpipe = function spullpipe(otherQ, el, cb) {
  reval(this._redis, safepullpipe_filename, [this._key, otherQ._key], [el], cb);
}

// Pop an element out of a queue and put it in another queue atomically
// Return the element being popped and pushed or nil
Q.prototype.poppipe = function poppipe(otherQ, cb) {
  this._redis.rpoplpush(this._key, otherQ._key, cb);
};

// Block and Pop an element out of a queue and put it in another queue atomically
// Return the element being popped and pushed
Q.prototype.bpoppipe = function bpoppipe(otherQ, timeout, cb) {
  if (cb === undefined) {
    cb = timeout;
    timeout = 0;
  }

  this._redis.brpoplpush(this._key, otherQ._key, timeout, cb);
};

// Clear the queue of elements
Q.prototype.clear = function clear(cb) {
  this._redis.del(this._key, cb);
};

// List all the elements in the queue
Q.prototype.list = function list(cb) {
  this._redis.lrange(this._key, 0, -1, cb);
};


// Create an event emitter for elements as we pip them
// options:
//  - max_out: Maximum callbacks allowed out at once (default 0 ~ infinity)
/*
    var listener = q.poplisten(otherQ, {max_out: 10})
      .on('message', function (msg, done) {
        // do something...
        done();
      })
      .on('error', function (err) {...});
    });

    listener.end();
*/
Q.prototype.poplisten = function poplisten(options) {
  if (this._listened) {
    throw new Error('You can\'t call a listen function more than once. Its not prudent.');
  }
  this._listened = true;

  options = options || {};

  var max_out = options.max_out || 0,
    clone = this.clone(),
    blockfunc = function (callback) { clone.bpop(0, callback); };

  var listener = new Listener(blockfunc, max_out);

  listener.once('end', function () {
    this._listened = false;
    clone._redis.end();
  });

  return listener.start();
};

// Create an event emitter for elements as we poppipe them
// otherQ - the otherQ parameter to poppipe
// options:
//  - max_out: Maximum callbacks allowed out at once (default 0 ~ infinity)
/*
    var listener = q.poppipelisten(otherQ, {max_out: 10})
      .on('message', function (msg, done) {
        // do something...
        done();
      })
      .on('error', function (err) {...});
    });

    listener.end();
*/
Q.prototype.poppipelisten = function poppipelisten(otherQ, options) {
  if (this._listened) {
    throw new Error('You can\'t call a listen function more than once. Its not prudent.');
  }
  this._listened = true;

  options = options || {};

  var max_out = options.max_out || 0,
    clone = this.clone(),
    blockfunc = function (callback) { clone.bpoppipe(otherQ, 0, callback); };

  var listener = new Listener(blockfunc, max_out);

  listener.once('end', function () {
    this._listened = false;
    clone._redis.end();
  });

  return listener.start();
};


exports.Q = exports.Queue = Q;

File: simpleq/simpleq.go
// Package simpleq provides a super-simple queue backed by Redis
package simpleq

import (
	"github.com/Rafflecopter/golang-simpleq/scripts"
	"github.com/garyburd/redigo/redis"
)

// A super simple redis-backed queue
type SimpleQ struct {
	conn     redis.Conn
	key      string
	listener *Listener
}

// Create a simpleq
func New(conn redis.Conn, key string) *SimpleQ {
	return &SimpleQ{
		conn: conn,
		key:  key,
	}
}

// End this queue
func (q *SimpleQ) Close() error {
	err := q.conn.Close()
	if err != nil {
		return err
	}
	if q.listener != nil {
		return q.listener.Close()
	}
	return nil
}

// Push an element onto the queue
func (q *SimpleQ) Push(el []byte) (length int64, err error) {
	return redis.Int64(q.conn.Do("LPUSH", q.key, el))
}

// Pop an element off the queue
func (q *SimpleQ) Pop() (el []byte, err error) {
	return redis.Bytes(q.conn.Do("RPOP", q.key))
}

// Block and Pop an element off the queue
// Use timeout_secs = 0 to block indefinitely
// On timeout, this DOES return an error because redigo does.
func (q *SimpleQ) BPop(timeout_secs int) (el []byte, err error) {
	res, err := redis.Values(q.conn.Do("BRPOP", q.key, timeout_secs))

	if len(res) == 2 {
		if bres, ok := res[1].([]byte); ok {
			return bres, err
		}
	}
	return nil, err
}

// Pull an element out the queue (oldest if more than one)
func (q *SimpleQ) Pull(el []byte) (nRemoved int64, err error) {
	return redis.Int64(q.conn.Do("LREM", q.key, -1, el))
}

// Pull an element out of the queue and push it onto another atomically
// Note: This will push the element regardless of the return value from pull
func (q *SimpleQ) PullPipe(q2 *SimpleQ, el []byte) (lengthQ2 int64, err error) {
	c := q.conn

	c.Send("MULTI")
	c.Send("LREM", q.key, -1, el)
	c.Send("LPUSH", q2.key, el)
	res, err := redis.Values(c.Do("EXEC"))

	if len(res) == 2 {
		if ires, ok := res[1].(int64); ok {
			return ires, err
		}
	}

	return 0, err
}

// Safely pull an element out of the queue and push it onto another atomically
// Returns 0 for non-existance in first queue, or length of second queue
func (q *SimpleQ) SPullPipe(q2 *SimpleQ, el []byte) (result int64, err error) {
	return redis.Int64(scripts.SafePullPipe.Do(q.conn, q.key, q2.key, el))
}

// Pop an element out of a queue and put it in another queue atomically
func (q *SimpleQ) PopPipe(q2 *SimpleQ) (el []byte, err error) {
	return redis.Bytes(q.conn.Do("RPOPLPUSH", q.key, q2.key))
}

// Block and Pop an element out of a queue and put it in another queue atomically
// On timeout, this doesn't return an error because redigo doesn't.
func (q *SimpleQ) BPopPipe(q2 *SimpleQ, timeout_secs int) (el []byte, err error) {
	return redis.Bytes(q.conn.Do("BRPOPLPUSH", q.key, q2.key, timeout_secs))
}

// Clear the queue of elements
func (q *SimpleQ) Clear() (nRemoved int64, err error) {
	return redis.Int64(q.conn.Do("DEL", q.key))
}

// List the elements in the queue
func (q *SimpleQ) List() (elements [][]byte, err error) {
	res, err := redis.Values(q.conn.Do("LRANGE", q.key, 0, -1))
	if err != nil {
		return nil, err
	}

	elements = make([][]byte, len(res))
	for i, el := range res {
		elements[i], _ = el.([]byte)
	}
	return elements, nil
}

// Create a listener that calls Pop
func (q *SimpleQ) PopListen(conn redis.Conn) *Listener {
	return q.PopPipeListen(conn, nil)
}

// Create a listener that calls PopPipe
func (q *SimpleQ) PopPipeListen(conn redis.Conn, q2 *SimpleQ) *Listener {
	if q.listener != nil {
		panic("SimpleQ can only have one listener")
	}

	q.listener = NewListener(conn, q, q2)

	go func() {
		<-q.listener.ended
		q.listener = nil
	}()

	return q.listener
}
